components/script/dom/promisedebugging.rs

```diff
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use dom::bindings::inheritance::Castable;
+use dom::bindings::root::DomRoot;
+use dom::event::{Event, EventBubbles, EventCancelable};
+use dom::eventtarget::EventTarget;
+use dom::globalscope::GlobalScope;
+use dom::promise::Promise;
+use dom::promiserejectionevent::PromiseRejectionEvent;
+use js::jsapi::{Heap, JSObject, HandleObject as RawHandleObject};
+use js::rust::wrappers::GetPromiseResult;
+use std::cell::Cell;
+
+pub struct PromiseDebugging {
+  pub id_predix: Cell<String>,
+}
+
+impl PromiseDebugging {
+    #[allow(unsafe_code)]
+    pub fn AddUncaughtRejection(global: DomRoot<GlobalScope>, promise: RawHandleObject) {
+        global.set_uncaught_rejections(promise);
+
+        unsafe {
+            PromiseDebugging::flush_uncaught_rejections_internal(global);
+        }
+    }
+
+    #[allow(unsafe_code)]
+    pub unsafe fn flush_uncaught_rejections_internal(global: DomRoot<GlobalScope>) {
```

Comment:
@jdm cc @KiChjang 

For this comment, I still doesn't implement it yet due to below question:

I tried to figure out how to call `notify_about_rejected_promises` inside `MicrotaskQueue::checkpoint`.
Due to `GlobalScope` as an argument in `notify_about_rejected_promises`, we'll need to pass a `GlobalScope` from `MicrotaskQueue::checkpoint` but it seems that we don't have a way to get `GlobalScope` in `MicrotaskQueue`. So, maybe we can only pass a `GlobalScope` to it.

Ex.
```rust
pub fn checkpoint<F>(&self, global: GlobalScope, target_provider: F)
    where F: Fn(PipelineId) -> Option<DomRoot<GlobalScope>>
{
```

However, it will have an issue that we'll call `MicrotaskQueue::checkpoint` in two places:

One is in `ScriptThread`:
https://github.com/servo/servo/blob/d23bc4f1a4e3e56290f823b0466f7ca6905f8ccd/components/script/script_thread.rs#L2653-L2655

and the other is in `GlobalScope`:
https://github.com/servo/servo/blob/d23bc4f1a4e3e56290f823b0466f7ca6905f8ccd/components/script/dom/globalscope.rs#L521-L524

We must be able to get the `globalscope` inside `GlobalScope` module but we don't have `globalscope` under every `ScriptThread`.

How can we have `GlobalScope` for both caller places and pass it to `notify_about_rejected_promises`?
Do you have any advises?

Reply:
While investigating this question, I discovered and filed #20908. Since [step 3](https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model:notify-about-rejected-promises) says we need to notify about rejected promises for each global associated with the script thread, I propose that the `checkpoint` method should accept a `Vec<DomRoot<GlobalScope>>` argument and iterate over it to notify all of the contained globals. For ScriptThread, this can be created by [iterating](https://github.com/servo/servo/blob/f6762d54e27eec8b65e88d062694b6b46b5bfea7/components/script/script_thread.rs#L370) over all known documents and getting their global. For GlobalScope, it can be a vector consisting of the self object.

Reply:
Passing an vector of `GlobalScope` sounds reasonable to me and I will try it.

But, my only concern to the solution is that, in script thread, if there's no any global in it, then it won't trigger `notify_about_rejected_promises` method from `checkpoint` method, is this reasonable?
Or, that said, after having a solution for it like #20908 mentioned; then, we will have corresponding global in each `checkpoint`.

Reply:
I can't think of a way to have a script thread with no globals while still having rejected promises that require notification.

Reply:
Indeed. While having rejected promises, it should have its own globals.
I will try to use the vector way to implement it, thanks!